### DISCUSSION TITLE
Extra token on optional rule

### DISCUSSION THREAD

**Discussion by alemuller** [tree-sitter/tree-sitter#4703] [Q&A] [answered]
I'm trying to parse lists with elements delimited by white spaces.

``` js
const NL = repeat1(/[\n]/);
const WS = repeat1(/[\t ]/);

module.exports = grammar({
    name: 'test',

    extras: $ => [ WS, NL ],

    rules: {

        source_file: $ => repeat(
            $._stmt,
        ),

        _stmt: $ => seq(
            $.list, ';'
        ),

        list: $ => seq(
            $.sel,
            repeat(seq(WS, $.sel))
        ),

        sel: $ => seq(
            optional('.'),
            $.id,
            repeat(seq('.', $.id))
        ),

        id: $ => /[a-zA-Z]+/

    }
});
```

Test file:
```
.foo;
 foo;
.foo.bar;
 foo.bar;
 foo ;
```
The last line of test file has a white space before the semi colon.

```
$ tree-sitter --version
tree-sitter 0.19.4 (76862e281b3aa5d93b471bfdf903527fec4f9b29)
$ tree-sitter parse test.test
(source_file [0, 0] - [5, 0]
  (list [0, 0] - [0, 4]
    (sel [0, 0] - [0, 4]
      (id [0, 1] - [0, 4])))
  (list [1, 1] - [1, 4]
    (sel [1, 1] - [1, 4]
      (id [1, 1] - [1, 4])))
  (list [2, 0] - [2, 8]
    (sel [2, 0] - [2, 8]
      (id [2, 1] - [2, 4])
      (id [2, 5] - [2, 8])))
  (list [3, 1] - [3, 8]
    (sel [3, 1] - [3, 8]
      (id [3, 1] - [3, 4])
      (id [3, 5] - [3, 8])))
  (list [4, 1] - [4, 4]
    (sel [4, 1] - [4, 4]
      (id [4, 1] - [4, 4])))
  (ERROR [4, 4] - [4, 5]))
test.test	0 ms	(ERROR [4, 4] - [4, 5])
```

I tried adding `$.list` to `conflicts` to avoid the error but it had no effect.

**Comment by ahelwer** [comment, marked as answer]
@alemuller what is your desired behavior? The grammar seems to be behaving appropriately; on the final line, it finds a `WS` token which is not followed by a `sel` token before a `;`.

Assuming you want your grammar to ignore `WS` when it's directly followed by a `;`, you'd have to change your `_stmt` definition to:
```
_stmt: $ => seq($.list, optional(WS), ';'),
```
which of course generates a LR(1) parser conflict:
```
sel  •  '_stmt_token1'  …
```
Since when the parser finds a `WS` token within the context of a `list`, it can't tell (without looking ahead further) whether this `WS` token precedes another `sel` token or it precedes a `;` token. So you have to add a conflict for `list` to trigger GLR parsing:
```
conflicts: $ => [
  [$.list]
],
```
but then we run into an issue with the associativity of the `optional(WS)` we added above; from your definitions of `NL` and `WS`, `repeat1` allows arbitrary whitespace between its repeated tokens, which are themselves whitespace, so they are ambiguous. What you probably want to do is write:
```
const NL = repeat1(token.immediate(/[\n\r]/));
const WS = repeat1(token.immediate(/[\t ]/));
```
or equivalently:
```
const NL = /[\n\r]+/
const WS = /[\t ]+/
```
(I also added a `\r` to your `NL` definition to support Windows)

This will parse your example file without error.

Incidentally I don't think your `extras` definition does anything, since tree-sitter uses context-aware parsing. It will only look for whitespace when your grammar says it expects a `WS` or `NL` token.

**Comment by alemuller** [comment]
I was expecting tree-sitter to check for conflicts when extra tokens are used in the grammar.

Thanks for the comprehensive reply.

**Comment by ahelwer** [comment]
@alemuller I guess the way to reason about it is that when the parser sees the token sequence `sel    WS`, it will match the `WS` token with the `WS` expected in the `list` definition, which takes higher priority over ignoring it (as is the default). I'm not sure where this exactly falls on tree-sitter's [conflicting tokens rules](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens); maybe ignoring whitespace (or whatever tokens are defined in `extras`, which by default is whitespace) is an undocumented sixth-priority rule?
